// @react-three/drei/core/Effects.js
import _extends from '@babel/runtime/helpers/esm/extends';
import * as React from 'react';
import { WebGLRenderTarget, RGBAFormat } from 'three';
import { extend, useThree, useFrame } from '@react-three/fiber';
import { EffectComposer, RenderPass, ShaderPass, GammaCorrectionShader } from 'three-stdlib';
import mergeRefs from 'react-merge-refs';

extend({
  EffectComposer,
  RenderPass,
  ShaderPass
});
const isWebGL2Available = () => {
  try {
    var canvas = document.createElement('canvas');
    return !!(window.WebGL2RenderingContext && canvas.getContext('webgl2'));
  } catch (e) {
    return false;
  }
};
const Effects = /*#__PURE__*/React.forwardRef(({
  children,
  multisamping = 8,
  renderIndex = 1,
  disableGamma = false,
  disableRenderPass = false,
  ...props
}, ref) => {
  const composer = React.useRef();
  const {
    scene,
    camera,
    gl,
    size,
    viewport
  } = useThree();
  const [target] = React.useState(() => {
    const t = new WebGLRenderTarget(size.width, size.height, {
      format: RGBAFormat,
      encoding: gl.outputEncoding
    });
    t.samples = multisamping;
    return t;
  });
  React.useEffect(() => {
    var _composer$current, _composer$current2;

    (_composer$current = composer.current) == null ? void 0 : _composer$current.setSize(size.width, size.height);
    (_composer$current2 = composer.current) == null ? void 0 : _composer$current2.setPixelRatio(viewport.dpr);
  }, [gl, size, viewport.dpr]);
  useFrame(() => {
    var _composer$current3;

    return (_composer$current3 = composer.current) == null ? void 0 : _composer$current3.render();
  }, renderIndex);
  const passes = [];
  if (!disableRenderPass) passes.push( /*#__PURE__*/React.createElement("renderPass", {
    key: "renderpass",
    args: [scene, camera]
  }));
  if (!disableGamma) passes.push( /*#__PURE__*/React.createElement("shaderPass", {
    key: "gammapass",
    args: [GammaCorrectionShader]
  }));
  React.Children.forEach(children, (el, index) => passes.push( /*#__PURE__*/React.cloneElement(el, {
    key: index,
    attach: `passes-${index}`
  })));
  return /*#__PURE__*/React.createElement("effectComposer", _extends({
    ref: mergeRefs([ref, composer]),
    args: [gl, target]
  }, props), passes);
});

export { Effects, isWebGL2Available };
